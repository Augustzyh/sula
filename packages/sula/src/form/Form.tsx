import React from 'react';
import { Form as AForm, Spin } from 'antd';
import omit from 'lodash/omit';
import { FormProps as AFormProps } from 'antd/lib/form';
import {FieldGroupProps} from './FieldGroup';
import FieldGroupContext, { rootGroupName, HOOK_MARK } from './FieldGroupContext';
import { getItemLayout } from './utils/layoutUtil';
import { NormalizedItemLayout } from './FieldGroup';
import { FieldGroup } from '.';
import useFormContext from './useFormContext';
import FormDependency from './dependency';
import { triggerActionPlugin } from '../rope/triggerPlugin';
import MediaQueries from './MediaQueries';
import { RequestConfig } from '../types/request';
import { RenderPlugin, FieldPlugin } from '../types/plugin';

export interface FormProps extends Omit<AFormProps, 'children'>, Omit<FieldGroupProps, 'name' | 'initialVisible' | 'dependency'> {
  remoteValues?: RequestConfig;
  ctxGetter?: () => Record<string, any>;
}

const Form: React.FC<FormProps> = (props, ref) => {
  const [formInstance] = AForm.useForm();

  const [context] = useFormContext(formInstance);

  React.useImperativeHandle(ref, context.getForm);

  const [loading, setLoading] = React.useState(false);

  const { saveFormProps, saveFormDependency, cascade, getCtx } = context.getInternalHooks(
    HOOK_MARK,
  );

  const formDependencyRef = React.useRef(new FormDependency());

  saveFormDependency(formDependencyRef.current);

  saveFormProps(props);

  const {
    layout = 'horizontal',
    itemLayout,
    size = 'middle',
    mode,
    remoteValues,
    initialValues,
    container,
    fields,
    children,
    actionsRender,
    actionsPosition,
  } = props;

  React.useEffect(() => {
    const ctx = getCtx();
    if (initialValues) {
      ctx.form.setFieldsValue(initialValues);
    }

    if (mode !== 'create' && remoteValues && remoteValues.init !== false) {
      setLoading(true);
      triggerActionPlugin(ctx, remoteValues)
        .then((fieldsValue: any) => {
          ctx.form.setFieldsValue(fieldsValue);
          setLoading(false);
        })
        .catch(() => {
          setLoading(false);
        });
    }
  }, []);

  const finalChildren = fields ? (
    <FieldGroup
      fields={fields}
      container={container}
      actionsRender={actionsRender}
      actionsPosition={actionsPosition}
    />
  ) : (
    children
  );

  const formProps = omit(props, [
    'itemLayout',
    'mode',
    'remoteValues',
    'container',
    'fields',
    'children',
    'actionsRender',
    'actionsPosition',
    'ctxGetter',
  ]);

  const originValuesChange = formProps.onValuesChange;

  formProps.onValuesChange = (changedValue, allValues) => {
    cascade(changedValue);
    if (originValuesChange) {
      originValuesChange(changedValue, allValues);
    }
  };

  return (
    <MediaQueries>
      {(matchedPoint) => {
        const normalizedItemLayout: NormalizedItemLayout = getItemLayout(
          itemLayout,
          layout,
          matchedPoint,
        );
        const fieldGroupContext = {
          formContext: context,
          layout,
          itemLayout: normalizedItemLayout,
          parentGroupName: rootGroupName,
          size,
          matchedPoint,
        };
        const wrapperChildren = (
          <FieldGroupContext.Provider value={fieldGroupContext}>
            {finalChildren}
          </FieldGroupContext.Provider>
        );

        return (
          <Spin spinning={loading}>
            <AForm
              {...formProps}
              wrapperCol={normalizedItemLayout.wrapperCol}
              labelCol={normalizedItemLayout.labelCol}
              children={wrapperChildren}
              form={formInstance}
            />
          </Spin>
        );
      }}
    </MediaQueries>
  );
};

export default Form;
