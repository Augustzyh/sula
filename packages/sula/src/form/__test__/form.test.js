import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Form from '..';
import { delay } from '../../__tests__/common';

describe('form', () => {
  describe('form props', () => {
    it('onValueChange cascade', async () => {
      let formRef;
      const onValuesChange = jest.fn();
      const form = (
        <Form
          ref={(ref) => {
            formRef = ref;
          }}
          onValuesChange={onValuesChange}
          fields={[
            { name: 'input', label: 'input', field: 'input' },
            {
              name: 'input2',
              label: 'inpt2',
              field: 'input',
              dependency: {
                value: {
                  relates: ['input'],
                  inputs: [['a']],
                  output: 'aaa',
                  defaultOutput: 'xxx',
                },
              },
            },
          ]}
        />
      );

      const wrapper = mount(form);
      wrapper
        .find('input')
        .at(0)
        .simulate('change', { target: { value: 'a' } });
      await wrapper.update();
      expect(formRef.getFieldsValue()).toEqual({ input: 'a', input2: 'aaa' });
      expect(onValuesChange).toHaveBeenCalled();
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('remoteValues', async () => {
      let formRef;
      const form = (
        <Form
          ref={(ref) => {
            formRef = ref;
          }}
          remoteValues={{
            url: '/manage.json',
            method: 'post',
          }}
          fields={[
            { name: 'input', label: 'input', field: 'input' },
            {
              name: 'input2',
              label: 'inpt2',
              field: 'input',
              dependency: {
                value: {
                  relates: ['input'],
                  inputs: [['a']],
                  output: 'aaa',
                  defaultOutput: 'xxx',
                },
              },
            },
          ]}
        />
      );

      const wrapper = mount(form);
      await delay(1000);
      expect(formRef.getFieldsValue()).toEqual({ input: 'a', input2: 'aaa' });
      wrapper
        .find('input')
        .at(0)
        .simulate('change', { target: { value: 'x' } });
      wrapper.update();
      expect(formRef.getFieldsValue()).toEqual({ input: 'x', input2: 'xxx' });
    });

    it('remoteValues Error', async () => {
      let formRef;
      const form = (
        <Form
          ref={(ref) => {
            formRef = ref;
          }}
          remoteValues={{
            url: '/error.json', // 模拟请求报错
            method: 'post',
          }}
          fields={[
            { name: 'input', label: 'input', field: 'input' },
            {
              name: 'input2',
              label: 'inpt2',
              field: 'input',
              dependency: {
                value: {
                  relates: ['input'],
                  inputs: [['a']],
                  output: 'aaa',
                  defaultOutput: 'xxx',
                },
              },
            },
          ]}
        />
      );

      const wrapper = mount(form);
      await delay(1000);
      expect(formRef.getFieldsValue()).toEqual({ input: undefined, input2: undefined });
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('children container', () => {
    it('children container', () => {
      let formRef;
      const form = (
        <Form
          ref={(ref) => {
            formRef = ref;
          }}
          fields={[
            {
              label: 'Inputgroup',
              childrenContainer: {
                type: 'inputgroup',
                props: {
                  compact: true,
                },
              },
              children: [
                {
                  name: 'input1',
                  field: {
                    type: 'input',
                    props: {
                      style: { width: '50%' },
                    },
                  },
                },
                {
                  name: 'input2',
                  field: {
                    type: 'input',
                    props: {
                      style: { width: '50%' },
                    },
                  },
                },
              ],
            },
            {
              label: 'childrenbtn',
              childrenContainer: 'div',
              children: [
                {
                  label: 'btn1',
                  render: {
                    type: 'button',
                    props: {
                      children: 'button1',
                    },
                  },
                },
                {
                  label: 'btn2',
                  render: {
                    type: 'button',
                    props: {
                      children: 'button2',
                    },
                  },
                },
              ],
            },
          ]}
        />
      );

      expect(renderer.create(form).toJSON()).toMatchSnapshot();
    });
  });
});
