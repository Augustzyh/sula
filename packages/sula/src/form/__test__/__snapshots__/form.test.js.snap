// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form children container children container 1`] = `
<form
  class="ant-form ant-form-horizontal"
>
  <div
    childrencontainer="[object Object]"
    class="ant-row ant-form-item"
  >
    <div
      class="ant-col ant-col-8 ant-form-item-label"
    >
      <label
        class=""
        title="Inputgroup"
      >
        Inputgroup
      </label>
    </div>
    <div
      class="ant-col ant-col-12 ant-form-item-control"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <span
            class="ant-input-group ant-input-group-compact"
          >
            <div
              class="ant-row ant-form-item"
            >
              <div
                class="ant-col ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <input
                      class="ant-input"
                      id="input1"
                      style="width: 50%;"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="ant-row ant-form-item"
            >
              <div
                class="ant-col ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <input
                      class="ant-input"
                      id="input2"
                      style="width: 50%;"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    childrencontainer="div"
    class="ant-row ant-form-item"
  >
    <div
      class="ant-col ant-col-8 ant-form-item-label"
    >
      <label
        class=""
        title="childrenbtn"
      >
        childrenbtn
      </label>
    </div>
    <div
      class="ant-col ant-col-12 ant-form-item-control"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <div>
            <div
              class="ant-row ant-form-item"
            >
              <div
                class="ant-col ant-form-item-label"
              >
                <label
                  class=""
                  title="btn1"
                >
                  btn1
                </label>
              </div>
              <div
                class="ant-col ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <button
                      class="ant-btn"
                      type="button"
                    >
                      <span>
                        button1
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="ant-row ant-form-item"
            >
              <div
                class="ant-col ant-form-item-label"
              >
                <label
                  class=""
                  title="btn2"
                >
                  btn2
                </label>
              </div>
              <div
                class="ant-col ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <button
                      class="ant-btn"
                      type="button"
                    >
                      <span>
                        button2
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</form>
`;

exports[`form form props onValueChange cascade 1`] = `
<form
  class="ant-form ant-form-horizontal"
>
  <div
    class="ant-row ant-form-item ant-form-item-has-success"
  >
    <div
      class="ant-col ant-col-8 ant-form-item-label"
    >
      <label
        class=""
        for="input"
        title="input"
      >
        input
      </label>
    </div>
    <div
      class="ant-col ant-col-12 ant-form-item-control"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <input
            class="ant-input"
            id="input"
            type="text"
            value="a"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="ant-row ant-form-item"
    dependency="[object Object]"
  >
    <div
      class="ant-col ant-col-8 ant-form-item-label"
    >
      <label
        class=""
        for="input2"
        title="inpt2"
      >
        inpt2
      </label>
    </div>
    <div
      class="ant-col ant-col-12 ant-form-item-control"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <input
            class="ant-input"
            id="input2"
            type="text"
            value="aaa"
          />
        </div>
      </div>
    </div>
  </div>
</form>
`;

exports[`form form props remoteValues Error 1`] = `
<form
  class="ant-form ant-form-horizontal"
>
  <div
    class="ant-row ant-form-item"
  >
    <div
      class="ant-col ant-col-8 ant-form-item-label"
    >
      <label
        class=""
        for="input"
        title="input"
      >
        input
      </label>
    </div>
    <div
      class="ant-col ant-col-12 ant-form-item-control"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <input
            class="ant-input"
            id="input"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="ant-row ant-form-item"
    dependency="[object Object]"
  >
    <div
      class="ant-col ant-col-8 ant-form-item-label"
    >
      <label
        class=""
        for="input2"
        title="inpt2"
      >
        inpt2
      </label>
    </div>
    <div
      class="ant-col ant-col-12 ant-form-item-control"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <input
            class="ant-input"
            id="input2"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</form>
`;
